tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -317,35 +322,56 @@ initialize_batch(DecompressChunkState *state, TupleTableSlot *slot)
 		{
 			case COMPRESSED_COLUMN:
 			{
-				value = slot_getattr(slot, column->compressed_scan_attno, &isnull);
-				if (!isnull)
-				{
-					CompressedDataHeader *header = (CompressedDataHeader *) PG_DETOAST_DATUM(value);
+				column->compressed.iterator = NULL;
 
-					column->compressed.iterator =
-						tsl_get_decompression_iterator_init(header->compression_algorithm,
-															state->reverse)(PointerGetDatum(header),
-																			column->typid);
-				}
-				else
+				value = slot_getattr(compressed_slot, column->compressed_scan_attno, &isnull);
+
+				if (isnull)
+				{
+					/*
+					 * The column will have a default value for the entire batch,
+					 * set it now.
+					 */
 					column->compressed.iterator = NULL;
+					AttrNumber attr = AttrNumberGetAttrOffset(column->output_attno);
+					decompressed_slot->tts_values[attr] =
+						getmissingattr(decompressed_slot->tts_tupleDescriptor,
+									   attr + 1,
+									   &decompressed_slot->tts_isnull[attr]);
+					break;
+				}
+
+				CompressedDataHeader *header = (CompressedDataHeader *) PG_DETOAST_DATUM(value);
+
+				column->compressed.iterator =
+					tsl_get_decompression_iterator_init(header->compression_algorithm,
+														state->reverse)(PointerGetDatum(header),
+																		column->typid);
 
 				break;
 			}
 			case SEGMENTBY_COLUMN:
-				value = slot_getattr(slot, column->compressed_scan_attno, &isnull);
-				if (!isnull)
-					column->segmentby.value = value;
-				else
-					column->segmentby.value = (Datum) 0;
-
-				column->segmentby.isnull = isnull;
+			{
+				/*
+				 * A segmentby column is not going to change during one batch,
+				 * and our output tuples are read-only, so it's enough to only
+				 * save it once per batch, which we do here.
+				 */
+				AttrNumber attr = AttrNumberGetAttrOffset(column->output_attno);
+				decompressed_slot->tts_values[attr] =
+					slot_getattr(compressed_slot,
+								 column->compressed_scan_attno,
+								 &decompressed_slot->tts_isnull[attr]);
 				break;
+			}
 			case COUNT_COLUMN:
-				value = slot_getattr(slot, column->compressed_scan_attno, &isnull);
-				state->counter = DatumGetInt32(value);
+				value = slot_getattr(compressed_slot, column->compressed_scan_attno, &isnull);
 				/* count column should never be NULL */
 				Assert(!isnull);
+				int count_value = DatumGetInt32(value);
+				Assert(count_value > 0);
```

Comment:
Is it guaranteed that no empty batches exist (e.g., created due to a bug in a previous version)? Otherwise, the Assert should be changed.

```suggestion
				Assert(count_value >= 0);
```

Reply:
I think this would mean data corruption. I'll change it to an `ereport`. We don't control the data, so this is technically an environment error.

Reply:
I did some experiments and could not generate such chunks either. I think `ereport` is a good idea here.